core/src/main/scala/filodb/core/utils/ColumnFilterMap.scala

```diff
@@ -0,0 +1,164 @@
+                                          otherFilters: Iterable[(ColumnFilter, T)]) {
+    /**
+     * Returns an occupied Option iff at least one filter is matched.
+     * The Equals map is checked first before other filters are iterated.
+     */
+    def get(string: String): Option[T] = {
+      if (equalsFilters.contains(string)) {
+        return Some(equalsFilters(string))
+      }
+      val matchingFilter = otherFilters.find{ case (filter, _) => filter.filter.filterFunc(string) }
+      if (matchingFilter.isDefined) {
+        val (_, elt) = matchingFilter.get
+        return Some(elt)
+      }
+      None  // No match found.
+    }
+  }
+
+  // Stores a map for each possible sequence of labels.
+  // Keys are found deterministically from each set of ColumnFilters to store.
+  val labelSeqToMap = new mutable.HashMap[Seq[String], mutable.Map[Seq[String], FilterToElementMap]]
+
+  filterElementPairs.foreach{ case (filters, _) =>
+    require(filters.count(!_.filter.isInstanceOf[Filter.Equals]) <= 1,
+      "at most one filter can be non-Equals; filters=" + filters)
+  }
+
+  filterElementPairs.map{ case (filters, elt) =>
+    // Sort the filters to determine the appropriate labelSeqToMap key.
+    // Sort alphabetically, then move any non-Equals filters to the end.
+    val sortedFilters = filters.toSeq
+      .sortBy(_.column)
+      .sortWith { case (leftFilter, rightFilter) =>
+        leftFilter.filter.isInstanceOf[Filter.Equals] &&
+          !rightFilter.filter.isInstanceOf[Filter.Equals]
+      }
+      (sortedFilters, elt)
+  }.groupBy{ case (sortedFilters, _) =>
+    // Group each set of filters by:
+    //   - its sequence of label names (determines the map it will be assigned to)
+    //   - its sequence of label values (determines the key for storage within the assigned map)
+    val labelNames = sortedFilters.map(_.column)
+    val prefixValues = sortedFilters
+      .dropRight(1)
+      .map(_.filter.asInstanceOf[Filter.Equals].value.toString)
+    (labelNames, prefixValues)
+  }.foreach{ case ((labelNames, prefixValues), sortedFiltersToElements) =>
+    // All (filters,element) pairs here will be grouped at the same path in the same map.
+    // For some efficiency, we store all Equals filters as part of a map. All non-Equals
+    //   filters are stored in a separate list (to be iterated sequentially).
+    val equalsFilterMap = new mutable.HashMap[String, T]
+    val otherFilters = new ArrayBuffer[(ColumnFilter, T)]
+    sortedFiltersToElements.foreach { case (filters, element) =>
+      filters.last.filter match {
+        case Filter.Equals(value) => equalsFilterMap(value.toString) = element
+        case _ => otherFilters.append((filters.last, element))
+      }
+    }
+    val filterToElementMap = FilterToElementMap(equalsFilterMap.toMap, otherFilters)
+    labelSeqToMap
+      .getOrElseUpdate(labelNames, new mutable.HashMap[Seq[String], FilterToElementMap]())
+      .put(prefixValues, filterToElementMap)
```

Comment:
If I understand correctly,
```
filters = (l1=v1, l2=v2, l3=~"v3.*") 
will be converted to
Map([l1,l2,l3] => Map([v1, v2] => FilterToElementMap(filters)))
```
Do we have to store every filter into FilterToElementMap or just the last filter?

Reply:
Super close -- the last line should be:
```
// only the last filter is added to the map
Map([l1,l2,l3] => Map([v1, v2] => FilterToElementMap(filters.last)))
```

I'll steal this example, expand on it a bit, and add it to the docs.

Reply:
That makes sense. 
I have a question regarding this case. 
```
filters = ((l1=v1),  (l1=v2))
```
In this case, the filters will have the same key in the map?

Reply:
Yep, that's right 👍 
Snippet from some new docs:
```
  // As an example, consider a ColumnFilterMap composed of these entries:
...
  //     {l1="v1", l2="v2a"} -> F
  //     {l1="v1", l2="v2b"} -> G
  //     {l1="v1a", l2="v2a"} -> H
...
  //   The nested maps would be structured as:
...
  //     (l1, l2) -> {
  //         (v1) -> [l2="v2a":F, l2="v2b":G]
  //         (v1a) -> [l2="v2a":H]
  //     }
```

Reply:
Got it thank you for the explanation.